File: swift/Sources/EnvdoctorKit/Utils/Paths.swift
import Foundation

/// Render a path relative to the project root when possible, falling back to
/// the absolute path.
public func displayPath(_ rootDir: String, _ filePath: String) -> String {
    let rel = relativePath(rootDir, filePath)
    if rel.isEmpty || rel.hasPrefix("../") || rel == ".." || rel.hasPrefix("/") {
        return filePath
    }
    return rel
}

/// `path.relative`-style computation using POSIX separators.
public func relativePath(_ rootDir: String, _ filePath: String) -> String {
    let root = rootDir.hasSuffix("/") ? String(rootDir.dropLast()) : rootDir
    let file = filePath.hasSuffix("/") ? String(filePath.dropLast()) : filePath
    if file == root { return "" }
    let prefix = root + "/"
    if file.hasPrefix(prefix) {
        return String(file.dropFirst(prefix.count))
    }
    // General fallback: resolve ".." components lexically.
    let rootParts = root.split(separator: "/", omittingEmptySubsequences: true).map(String.init)
    let fileParts = file.split(separator: "/", omittingEmptySubsequences: true).map(String.init)
    var common = 0
    while common < rootParts.count, common < fileParts.count, rootParts[common] == fileParts[common] {
        common += 1
    }
    var rel: [String] = []
    rel.append(contentsOf: repeatElement("..", count: rootParts.count - common))
    rel.append(contentsOf: fileParts[common...])
    return rel.joined(separator: "/")
}

/// Join paths with platform separators, normalizing to forward slashes.
public func normalizePath(_ p: String) -> String {
    p.replacingOccurrences(of: "\\", with: "/")
}

File: dart/lib/src/utils/paths.dart
/// Path helpers shared by discovery and formatters.
library;

/// Path relative to root using forward slashes; null when not under root.
String? relativeTo(String root, String path) {
  final prefix = root.endsWith('/') ? root : '$root/';
  if (path.startsWith(prefix)) return path.substring(prefix.length);
  return path == root ? '' : null;
}

/// Render a path relative to the project root when possible, falling back to
/// the absolute path.
String displayPath(String rootDir, String filePath) {
  final rel = relativeTo(rootDir, filePath);
  if (rel != null && rel.isNotEmpty) return rel;
  return filePath;
}

/// Join with forward slashes (paths are already platform-separated on the
/// platforms envdoctor supports; this normalizes for output).
String normalizePath(String p) => p.replaceAll('\\', '/');

/// True for absolute paths: POSIX `/…`, Windows drive letters, or UNC.
bool isAbsolutePath(String p) =>
    p.startsWith('/') ||
    p.startsWith('\\\\') ||
    (p.length > 2 && RegExp(r'^[A-Za-z]:[\\/]').hasMatch(p));

String basename(String p) {
  final idx = p.lastIndexOf('/');
  return idx < 0 ? p : p.substring(idx + 1);
}

String extensionOf(String p) {
  final base = basename(p);
  final dot = base.lastIndexOf('.');
  if (dot <= 0) return '';
  return base.substring(dot + 1);
}

String dirname(String p) {
  final idx = p.lastIndexOf('/');
  if (idx < 0) return '';
  return idx == 0 ? '/' : p.substring(0, idx);
}

String joinPath(String a, String b) {
  if (a.endsWith('/')) return '$a$b';
  return '$a/$b';
}
